Submodule on second level fails to find itself in dev mode

### Describe the bug

My project consists of parent pom with common dependencies and several submodules. When I run dev mode inside submodule on the first level (folder project/submodule) it starts successfully. When I do the same for submodule on the second level (folder project/something/submodule) it fails with error `The following artifacts could not be resolved: ${GAV of the submodule}`. This used to work before Quarkus 3.7

### Expected behavior

Application starts in dev mode

### Actual behavior

```
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------< io.quarkus.ts.startstop:app-jakarta-rest-minimal >----------
[INFO] Building app-jakarta-rest-minimal 1.0.0-SNAPSHOT
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ app-jakarta-rest-minimal ---
[INFO] Deleting $EDITEDtest/app-jakarta-rest-minimal/target
[INFO] 
[INFO] --- quarkus:3.7.1:dev (default-cli) @ app-jakarta-rest-minimal ---
[INFO] Invoking formatter:2.22.0:format (default) @ app-jakarta-rest-minimal
[INFO] Processed 2 files in 518ms (Formatted: 0, Unchanged: 2, Failed: 0, Readonly: 0)
[INFO] Invoking impsort:1.8.0:sort (default) @ app-jakarta-rest-minimal
[INFO] Processed 2 files in 00:00.012 (Already Sorted: 2, Needed Sorting: 0)
[INFO] Invoking resources:3.3.1:resources (default-resources) @ app-jakarta-rest-minimal
[INFO] Copying 2 resources from src/main/resources to target/classes
[INFO] Invoking compiler:3.11.0:compile (default-compile) @ app-jakarta-rest-minimal
[INFO] Changes detected - recompiling the module! :source
[INFO] Compiling 2 source files with javac [debug target 17] to target/classes
[WARNING] system modules path not set in conjunction with -source 17
[INFO] Annotation processing is enabled because one or more processors were found
  on the class path. A future release of javac may disable annotation processing
  unless at least one processor is specified by name (-processor), or a search
  path is specified (--processor-path, --processor-module-path), or annotation
  processing is enabled explicitly (-proc:only, -proc:full).
  Use -Xlint:-options to suppress this message.
  Use -proc:none to disable annotation processing.
[INFO] Invoking resources:3.3.1:testResources (default-testResources) @ app-jakarta-rest-minimal
[INFO] skip non existing resourceDirectory /home/fedinsky/code/quarkus/refresh/quarkus-startstop/test/app-jakarta-rest-minimal/src/test/resources
[INFO] Invoking compiler:3.11.0:testCompile (default-testCompile) @ app-jakarta-rest-minimal
[INFO] No sources to compile
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.347 s
[INFO] Finished at: 2024-03-12T17:24:24+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:3.7.1:dev (default-cli) on project app-jakarta-rest-minimal: Failed to run: Failed to resolve artifact io.quarkus.ts.startstop:app-jakarta-rest-minimal:jar:1.0.0-SNAPSHOT: The following artifacts could not be resolved: io.quarkus.ts.startstop:app-jakarta-rest-minimal:jar:1.0.0-SNAPSHOT (absent): Could not find artifact io.quarkus.ts.startstop:app-jakarta-rest-minimal:jar:1.0.0-SNAPSHOT -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException

```

### How to Reproduce?

1. `git clone https://github.com/quarkus-qe/quarkus-startstop.git`
2. ` cd quarkus-startstop/`
3. `mkdir test`
4. `cp -r app-jakarta-rest-minimal/ test/` # to make sure they are identical
5. `cd test/app-jakarta-rest-minimal/`
6. `sed -i -e 's|>..<|>../..<|' pom.xml` # to have correct relative path
7. `mvn clean quarkus:dev -Dquarkus.version=3.7.1 # fails`
8. `mvn clean quarkus:dev -Dquarkus.version=3.6.4 # succeeds`

Running any of the commands above inside original `app-jakarta-rest-minimal/`always succeeds. Latest version (3.8.2) is also affected.

### Output of `uname -a` or `ver`

6.7.3-200.fc39.x86_64

### Output of `java -version`

Java version: 21.0.1, vendor: Eclipse Adoptium

### Quarkus version or git rev

3.7.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)

### Additional information

_No response_

--- a/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/WorkspaceLoader.java
+++ b/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/WorkspaceLoader.java
@@ -186,7 +186,6 @@ public class WorkspaceLoader implements WorkspaceModelResolver, WorkspaceReader 
         if (rawModule.model == null) {
             return;
         }
-        newModules.add(rawModule);
 
         var added = loadedModules.putIfAbsent(
                 new GAV(ModelUtils.getGroupId(rawModule.model), rawModule.model.getArtifactId(),
@@ -195,6 +194,8 @@ public class WorkspaceLoader implements WorkspaceModelResolver, WorkspaceReader 
         if (added != null) {
             return;
         }
+        newModules.add(rawModule);
+
         for (var module : rawModule.model.getModules()) {
             queueModule(rawModule.model.getProjectDirectory().toPath().resolve(module));
         }

--- a/independent-projects/bootstrap/maven-resolver/src/test/java/io/quarkus/bootstrap/workspace/test/LocalWorkspaceDiscoveryTest.java
+++ b/independent-projects/bootstrap/maven-resolver/src/test/java/io/quarkus/bootstrap/workspace/test/LocalWorkspaceDiscoveryTest.java
@@ -3,6 +3,7 @@
  */
 package io.quarkus.bootstrap.workspace.test;
 
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertNull;
@@ -20,7 +21,6 @@ import java.util.Properties;
 
 import org.apache.maven.model.Dependency;
 import org.apache.maven.model.Parent;
-import org.assertj.core.api.Assertions;
 import org.eclipse.aether.artifact.DefaultArtifact;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.AfterEach;
@@ -122,6 +122,30 @@ public class LocalWorkspaceDiscoveryTest {
     @AfterAll
     public static void cleanup() {
         IoUtils.recursiveDelete(workDir);
+    }
+
+    /**
+     * This test is making sure the current module isn't overridden by another module
+     * from the workspace that happens to have the same group and artifact IDs
+     *
+     * @throws Exception
+     */
+    @Test
+    public void workspaceWithDuplicateModuleGroupIdAndArtifactId() throws Exception {
+        final URL moduleUrl = Thread.currentThread().getContextClassLoader()
+                .getResource("duplicate-ga/test/case");
+        assertNotNull(moduleUrl);
+        final Path moduleDir = Path.of(moduleUrl.toURI());
+        assertNotNull(moduleUrl);
+
+        final LocalWorkspace ws = LocalProject.loadWorkspace(moduleDir).getWorkspace();
+
+        LocalProject project = ws.getProject("org.acme", "acme-lib");
+        assertNotNull(project);
+        assertThat(project.getDir()).isEqualTo(moduleDir);
+
+        assertNotNull(ws.getProject("org.acme", "acme-parent"));
+        assertEquals(2, ws.getProjects().size());
     }
 
     @Test
@@ -716,15 +740,15 @@ public class LocalWorkspaceDiscoveryTest {
         assertEquals(new File(rootPomUrl.toURI()), root);
 
         final WorkspaceModule wsModule = module1.toWorkspaceModule();
-        Assertions.assertThat(wsModule.getModuleDir()).isEqualTo(module1Dir.toFile());
-        Assertions.assertThat(wsModule.getBuildDir()).isEqualTo(module1Dir.resolve("target").toFile());
+        assertThat(wsModule.getModuleDir()).isEqualTo(module1Dir.toFile());
+        assertThat(wsModule.getBuildDir()).isEqualTo(module1Dir.resolve("target").toFile());
         SourceDir src = wsModule.getMainSources().getResourceDirs().iterator().next();
         PathTree sourceTree = src.getSourceTree();
-        Assertions.assertThat(sourceTree).isNotNull();
+        assertThat(sourceTree).isNotNull();
         Collection<Path> roots = sourceTree.getRoots();
-        Assertions.assertThat(roots).hasSize(1);
-        Assertions.assertThat(roots.iterator().next()).isEqualTo(module1Dir.resolve("build"));
-        Assertions.assertThat(src.getOutputDir()).isEqualTo(module1Dir.resolve("target/classes/META-INF/resources"));
+        assertThat(roots).hasSize(1);
+        assertThat(roots.iterator().next()).isEqualTo(module1Dir.resolve("build"));
+        assertThat(src.getOutputDir()).isEqualTo(module1Dir.resolve("target/classes/META-INF/resources"));
     }
 
     private void assertCompleteWorkspace(final LocalProject project) {

--- a/independent-projects/bootstrap/maven-resolver/src/test/resources/duplicate-ga/module/pom.xml
+++ b/independent-projects/bootstrap/maven-resolver/src/test/resources/duplicate-ga/module/pom.xml
@@ -1,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project>
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.acme</groupId>
+        <artifactId>acme-parent</artifactId>
+        <version>1.0-SNAPSHOT</version>
+    </parent>
+    <artifactId>acme-lib</artifactId>
+</project>

--- a/independent-projects/bootstrap/maven-resolver/src/test/resources/duplicate-ga/pom.xml
+++ b/independent-projects/bootstrap/maven-resolver/src/test/resources/duplicate-ga/pom.xml
@@ -1,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project>
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.acme</groupId>
+    <artifactId>acme-parent</artifactId>
+    <version>1.0-SNAPSHOT</version>
+    <packaging>pom</packaging>
+
+    <profiles>
+        <profile>
+            <id>extra-module</id>
+            <modules>
+                <module>module</module>
+            </modules>
+        </profile>
+    </profiles>
+</project>

--- a/independent-projects/bootstrap/maven-resolver/src/test/resources/duplicate-ga/test/case/pom.xml
+++ b/independent-projects/bootstrap/maven-resolver/src/test/resources/duplicate-ga/test/case/pom.xml
@@ -1,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project>
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.acme</groupId>
+        <artifactId>acme-parent</artifactId>
+        <version>1.0-SNAPSHOT</version>
+        <relativePath>../..</relativePath>
+    </parent>
+    <artifactId>acme-lib</artifactId>
+</project>

